### PR TITLE
Convert mod_mam to gen_hook

### PR DESCRIPTION
**Pull request by kamilwaz**
Convert mod_mam to gen_hook.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/148722) / small_tests / 9328d7bacb3a202ba3cfe6b5e4018e12385ade96
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/148722/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/148722/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-03_13.28.17/lib.mongooseim.logs/run.2022-11-03_13.28.18/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149082) / small_tests / 5afc98551944f06775bcd9ed89ef03b76d447788
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149082/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149082/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-04_13.56.38/lib.mongooseim.logs/run.2022-11-04_13.56.39/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149325) / small_tests / 781238a906072976f85ba2486092b3f50ee67e25
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149325/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149325/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-07_11.26.19/lib.mongooseim.logs/run.2022-11-07_11.26.20/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149355) / small_tests / b0de2c79ebd1c405cec63aa3fd2f03c744610b6f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149355/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149355/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-07_11.27.50/lib.mongooseim.logs/run.2022-11-07_11.27.51/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149479) / small_tests / 3da6535f941c549457333d3ac03c39c7a7841515
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149479/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149479/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-07_14.15.19/lib.mongooseim.logs/run.2022-11-07_14.15.19/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149505) / small_tests / 5e69c4caf9f7077bddcc3f73bcc9437ade8c682a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149505/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149505/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-07_14.47.49/lib.mongooseim.logs/run.2022-11-07_14.47.49/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149557) / small_tests / d9565e2213127fc80442791deab63bb5e4d02136
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149557/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149557/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-07_16.02.39/lib.mongooseim.logs/run.2022-11-07_16.02.39/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149608) / small_tests / 1be4e6243fe47b9226df72296f47add9db9d0baa
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149608/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149608/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-07_16.50.48/lib.mongooseim.logs/run.2022-11-07_16.50.49/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149862) / small_tests / b4a25be67e9804cf5189898283457dd8b2c58012
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149862/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149862/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-08_08.20.57/lib.mongooseim.logs/run.2022-11-08_08.20.57/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149920) / small_tests / 4a952df6ba726368d2ad1768df09021b0a9726f0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149920/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149920/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-08_08.55.45/lib.mongooseim.logs/run.2022-11-08_08.55.46/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/149944) / small_tests / 33517686729007d1d4c94acdf5865992d2d1ad9d
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/149944/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/149944/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-08_09.09.00/lib.mongooseim.logs/run.2022-11-08_09.09.00/suite.log.html)

File: src/mongoose_hooks.erl [hunk 1]
```diff
@@ -1204,93 +1219,106 @@ mam_retraction(HostType, RetractionInfo, Env) ->
...
 
 %%% @doc The `mam_muc_lookup_messages' hook is to retrieve archived
 %%% MUC messages for any given search parameters.
 -spec mam_muc_lookup_messages(HostType, Params) -> Result when
       HostType :: mongooseim:host_type(),
       Params :: map(),
       Result :: {ok, mod_mam:lookup_result()}.
-mam_muc_lookup_messages(HostType, Params) ->
+mam_muc_lookup_messages(HostType, Params0) ->
+    Params = #{params => Params0},
```
Comment (NelsonVides) [inline src/mongoose_hooks.erl:1321]: Same applies here for the `params` nesting.

File: src/mam/mod_mam_cassandra_prefs.erl [hunk 2]
```diff
@@ -100,24 +102,26 @@ get_behaviour(DefaultBehaviour, HostType, _UserID, LocJID, RemJID) ->
...
-            decode_behaviour(Behaviour)
+            {ok, decode_behaviour(Behaviour)}
     end.
 
 
--spec set_prefs(Result :: any(), HostType :: host_type(),
-                ArchiveID :: mod_mam:archive_id(), ArchiveJID :: jid:jid(),
-                DefaultMode :: mod_mam:archive_behaviour(),
-                AlwaysJIDs :: [jid:literal_jid()],
-                NeverJIDs :: [jid:literal_jid()]) -> any().
```
Comment (NelsonVides) [inline src/mam/mod_mam_cassandra_prefs.erl:111]: I'd be good if all these types were encoded and not lost. Perhaps an auxiliary type to declare in the same place the callback is declared, so that handlers can just use it.

File: src/mam/mod_mam_pm.erl [hunk 3]
```diff
@@ -673,8 +672,7 @@ hooks(HostType) ->
         {amp_determine_strategy, HostType, fun ?MODULE:determine_amp_strategy/3, #{}, 20},
         {sm_filter_offline_message, HostType, fun ?MODULE:sm_filter_offline_message/3, #{}, 50},
         {get_personal_data, HostType, fun ?MODULE:get_personal_data/3, #{}, 50}
-        | mongoose_metrics_mam_hooks:get_mam_hooks(HostType)
-    ].
```
Comment (NelsonVides) [inline src/mam/mod_mam_pm.erl:677]: Why this change? The previous version would technically be just as correct and also generate less intermediate garbage.

File: src/mam/mod_mam_pm.erl [hunk 4]
```diff
@@ -662,7 +662,6 @@ is_archivable_message(HostType, Dir, Packet) ->
     ArchiveChatMarkers = mod_mam_params:archive_chat_markers(?MODULE, HostType),
     erlang:apply(M, is_archivable_message, [?MODULE, Dir, Packet, ArchiveChatMarkers]).
 
--spec hooks(mongooseim:host_type()) -> gen_hook:hook_list().
```
Comment (NelsonVides) [inline src/mam/mod_mam_pm.erl:665]: Why removing the spec? 🤔

**Review by NelsonVides — COMMENTED** [review]
Just a few comments from an unfinished review.

File: src/mam/mod_mam_pm.erl (hunk 3 above)
Comment (kamilwaz) [reply src/mam/mod_mam_pm.erl:677]: Under the hood, the generated opcode would be the same. Dunno why I changed it, gonna revert it back.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/150401) / small_tests / 57220c8929f1bf17ce82612970cd318e70874d45
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/150401/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/150401/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-09_10.03.00/lib.mongooseim.logs/run.2022-11-09_10.03.01/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/150425) / small_tests / 170ee1b2a58e3b1d3d7f8e9a1eca97e0fad9d143
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/150425/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/150425/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-09_10.06.15/lib.mongooseim.logs/run.2022-11-09_10.06.16/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/150451) / small_tests / d4b602647e01e29929b356e8d684ba7bf4fa84e9
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/150451/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/150451/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-09_11.56.17/lib.mongooseim.logs/run.2022-11-09_11.56.17/suite.log.html)

File: src/mam/mod_mam_rdbms_arch.erl [hunk 5]
```diff
@@ -340,12 +338,13 @@ prepare_insert(Name, NumRows) ->
     ok.
 
 %% Removal logic
--spec remove_archive(Acc :: mongoose_acc:t(), HostType :: host_type(),
-                     ArcID :: mod_mam:archive_id(),
-                     RoomJID :: jid:jid()) -> mongoose_acc:t().
-remove_archive(Acc, HostType, ArcID, _ArcJID) ->
+-spec remove_archive(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: term(),
+    Params :: map(),
```
Comment (NelsonVides) [inline src/mam/mod_mam_rdbms_arch.erl:343]: All these params should also not lose type information, they could be like `#{archive_id := mod_mam:archive_id()}`. That is useful even if only when debugging this information, it already tells the reader what are we handling.

File: src/mam/mod_mam_riak_timed_arch_yz.erl [hunk 6]
```diff
@@ -96,40 +96,52 @@ mam_bucket_type(Host) ->
...
 %% RemJID - interlocutor's JID
 %% SrcJID - "Real" sender JID
-archive_message(_Result, Host, #{message_id := MessId,
-                                 local_jid := LocJID,
-                                 remote_jid := RemJID,
-                                 source_jid := SrcJID,
-                                 packet := Packet} = Params) ->
+-spec archive_message(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: ok,
+    Params :: map(),
```
Comment (NelsonVides) [inline src/mam/mod_mam_riak_timed_arch_yz.erl:101]: Likewise, there's a type defined for this that we could use.

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 7]
```diff
@@ -34,140 +34,140 @@
...
+        {mam_muc_remove_archive, Host, fun ?MODULE:mam_muc_remove_archive/3, #{}, 50},
+        {mam_muc_lookup_messages, Host, fun ?MODULE:mam_muc_lookup_messages/3, #{}, 100},
+        {mam_muc_flush_messages, Host, fun ?MODULE:mam_muc_flush_messages/3, #{}, 50}
     ].
 
 -spec mam_get_prefs(Acc, Params, Extra) -> {ok, Acc} when
-    Acc :: any(),
+    Acc :: mod_mam:preference() | {error, Reason :: term()},
     Params :: map(),
-    Extra :: gen_hook:extra().
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:63]: Why removing such type information? It is indeed a `gen_hook:extra()` here.

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 8]
```diff
@@ -34,140 +34,140 @@
...
-mam_get_prefs(Acc, _, #{host_type := HostType}) ->
-    mongoose_metrics:update(HostType, modMamPrefsGets, 1),
-    {ok, Acc}.
+    Extra :: map().
+mam_get_prefs(Result, _Params, #{host_type := Host}) ->
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:64]: Also these are all `HostType`s, not `Host`s, why the change?

File: src/mongoose_hooks.erl [hunk 9]
```diff
@@ -1063,125 +1063,138 @@ room_new_affiliations(Acc, Room, NewAffs, Version) ->
...
-    run_hook_for_host_type(mam_remove_archive, HookServer, ok,
+    run_hook_for_host_type(mam_remove_archive, HostType, ok,
                            ParamsWithLegacyArgs).
 
 %%% @doc The `mam_lookup_messages' hook is to retrieve
 %%% archived messages for given search parameters.
--spec mam_lookup_messages(HookServer, Params) -> Result when
-      HookServer :: jid:lserver(),
+-spec mam_lookup_messages(HostType, Params) -> Result when
+      HostType :: jid:lserver(),
```
Comment (NelsonVides) [inline src/mongoose_hooks.erl:1159]: If this is a `HostType`, the type name could also change to `mongooseim:host_type()`. Same applies to other hooks below.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/151528) / small_tests / 4a5ac6937e6d4f36116d0088cc1078a0d6433ab5
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/151528/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/151528/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-14_10.32.02/lib.mongooseim.logs/run.2022-11-14_10.32.02/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/151558) / small_tests / 9e72e3f52cff279701fd476123edd1c550d48a3a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/151558/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/151558/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-14_13.00.48/lib.mongooseim.logs/run.2022-11-14_13.00.48/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/151662) / small_tests / 6c4830d48fdeae184e66d7eb044de2a98ad6988f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/151662/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/151662/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-14_14.22.16/lib.mongooseim.logs/run.2022-11-14_14.22.16/suite.log.html)

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 10]
```diff
@@ -96,78 +97,77 @@ mam_lookup_messages(Result = {ok, {_TotalCount, _Offset, MessageRows}},
...
--spec mam_muc_lookup_messages(Result, Params, Extra) -> {ok, Result} when
-    Result :: {ok | error, mod_mam:lookup_result()},
-    Params :: map(),
-    Extra :: gen_hook:extra().
-mam_muc_lookup_messages(Result = {ok, {_TotalCount, _Offset, MessageRows}},
-                        _, #{host_type := HostType}) ->
+-spec mam_muc_lookup_messages(Acc, Params, Extra) -> {ok, Acc} when
+    Acc :: {ok, mod_mam:lookup_result()} | {error, term()},
+    Params :: mam_iq:lookup_params(),
+    Extra :: map().
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:149]:
```suggestion
    Extra :: gen_hook:extra().
```
And similarly in a few others in this module

**Review by NelsonVides — COMMENTED** [review]
Just a few very hidden typos and forgotten types 😄

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/151811) / small_tests / 5782c7cf46d296da6f449b79b85c2bb09b0832b8
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/151811/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/151811/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-15_08.45.45/lib.mongooseim.logs/run.2022-11-15_08.45.45/suite.log.html)

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 11]
```diff
@@ -96,78 +97,77 @@ mam_lookup_messages(Result = {ok, {_TotalCount, _Offset, MessageRows}},
...
+mam_archive_message(Result, _Params, #{host_type := HostType}) ->
     mongoose_metrics:update(HostType, modMamArchived, 1),
-    {ok, Acc}.
+    {ok, Result}.
 
 -spec mam_flush_messages(Acc, Params, Extra) -> {ok, Acc} when
-    Acc :: any(),
-    Params :: #{message_count := integer()},
+    Acc :: ok,
+    Params :: map(),
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:113]: Here some type knowledge was removed

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 12]
```diff
@@ -34,59 +34,60 @@
 %%-------------------
 
 %% @doc Here will be declared which hooks should be registered when mod_mam_pm is enabled.
--spec get_mam_hooks(_) -> gen_hook:hook_list().
-get_mam_hooks(HostType) ->
+-spec get_mam_hooks(mongooseim:host_type()) -> gen_hook:hook_list().
+get_mam_hooks(Host) ->
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:38]: All these varnames should stay as `HostType`s, here and in the function below as well.

**Review by NelsonVides — COMMENTED** [review]
Just two more comments 😄 
This PR has so many changes, mostly mechanical ones, that it gets quite confusing to review, sorry for not noticing these few more errors before 😅

**Review by NelsonVides — APPROVED** [review]
💪🏽 💪🏽

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/151840) / small_tests / f5402da67fe0bdd41002c413016f117ad195265e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/151840/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/151840/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-15_09.04.08/lib.mongooseim.logs/run.2022-11-15_09.04.08/suite.log.html)

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 13]
```diff
@@ -96,78 +97,77 @@ mam_lookup_messages(Result = {ok, {_TotalCount, _Offset, MessageRows}},
...
-mam_archive_message(Acc, _, #{host_type := HostType}) ->
+mam_archive_message(Result, _Params, #{host_type := HostType}) ->
     mongoose_metrics:update(HostType, modMamArchived, 1),
-    {ok, Acc}.
+    {ok, Result}.
 
 -spec mam_flush_messages(Acc, Params, Extra) -> {ok, Acc} when
-    Acc :: any(),
+    Acc :: ok,
     Params :: #{message_count := integer()},
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:113]: This turned into `count`, not `message_count` 😅

File: src/metrics/mongoose_metrics_mam_hooks.erl [hunk 14]
```diff
@@ -96,78 +97,77 @@ mam_lookup_messages(Result = {ok, {_TotalCount, _Offset, MessageRows}},
...
+mam_muc_archive_message(Result, _Params, #{host_type := HostType}) ->
     mongoose_metrics:update(HostType, modMucMamArchived, 1),
-    {ok, Acc}.
+    {ok, Result}.
 
 -spec mam_muc_flush_messages(Acc, Params, Extra) -> {ok, Acc} when
-    Acc :: any(),
-    Params :: #{message_count := integer()},
+    Acc :: ok,
+    Params :: map(),
```
Comment (NelsonVides) [inline src/metrics/mongoose_metrics_mam_hooks.erl:169]: Also here it is also `count := integer()`, no losing the type information better

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/151865) / small_tests / af2a2a49e4ef157bf39f4b5bd1e74ab21ddca99a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3841/151865/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3841/151865/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-15_09.36.11/lib.mongooseim.logs/run.2022-11-15_09.36.12/suite.log.html)
